File: src/examples/python/SampleAccount.py
#!/usr/bin/python

from gen_py.datahub.account import AccountService
from gen_py.datahub import DataHub
from gen_py.datahub.constants import *
from gen_py.datahub.account.constants import *
from thrift import Thrift
from thrift.protocol import TBinaryProtocol
from thrift.transport import THttpClient
from thrift.transport import TTransport

'''
@author: anant bhardwaj
@date: Oct 11, 2013

Sample Python client for DataHub Account Creation
'''



try:
  print "\n\nTrying to create and delete an account..."
  # This will initially thrown an exception since
  # You will need to register an application and assign
  # the variables below.
  # then, log in as the user, and visit this url to authorize the app:
  ERNAME
  # see the /developer/apps page to register an application.

  username="USERNAME"    
  password="PASSWORD"
  email="EMAIL"
  repo_name="REPO_NAME"       # the username of the account you are accessing
  app_id="APP_ID"             # your app's unique id
  app_token="APP_TOKEN"       # your app's unique token

  


  datahub_transport = THttpClient.THttpClient('http://datahub.csail.mit.edu/service')
  datahub_transport = TTransport.TBufferedTransport(datahub_transport)
  datahub_protocol = TBinaryProtocol.TBinaryProtocol(datahub_transport)
  datahub_client = DataHub.Client(datahub_protocol)
  print "Version: %s" %(datahub_client.get_version())

  account_transport = THttpClient.THttpClient('http://datahub.csail.mit.edu/service/account')
  account_transport = TTransport.TBufferedTransport(account_transport)
  account_protocol = TBinaryProtocol.TBinaryProtocol(account_transport)
  account_client = AccountService.Client(account_protocol)

  print "\nTrying to query a user's table..."

  # RECOMMENDED (How an app should read/write in a users' repo)
  # Connect to the user repository through app_id + app_token.
  # Don't expose their password
  con_params_app = ConnectionParams(app_id=app_id, app_token=app_token, repo_base=username)
  con_app = datahub_client.open_connection(con_params=con_params_app)

  createSQL ="CREATE TABLE "+repo_name+".app_table(content text)"
  insertSQL = "INSERT INTO "+repo_name+".app_table VALUES('Anant Bhardwaj')"
  selectSQL = "SELECT * FROM "+repo_name+".app_table"

  print datahub_client.list_repos(con=con_app)
  print datahub_client.list_tables(con=con_app, repo_name=repo_name)
  print datahub_client.execute_sql(con_app, createSQL, query_params=None)
  print datahub_client.list_tables(con=con_app, repo_name=repo_name)
  print datahub_client.execute_sql(con_app, insertSQL, query_params=None);
  print datahub_client.execute_sql(con_app, selectSQL, query_params=None);

except Exception, e:
  print 'Something went wrong : %s' % (e)
  print '\n Please make sure that you authorized the app at https://datahub.csail.mit.edu/permissions/apps/allow_access/' + app_id + '/' + repo_name

File: src/examples/java/src/SampleAccount.java
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.THttpClient;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import java.nio.ByteBuffer;
import datahub.*;
import datahub.account.*;
import datahub.account.AccountService;

/** 
 * Sample Java Client for DataHub
 * 
 * @author anantb
 * @date 11/07/2013
 * 
 */

public class SampleAccount {
  public static void main(String [] args) {
    try {
      // This will initially thrown an exception since
      // You will need to register an application and assign
      // the variables below.
      // then, log in as the user, and visit this url to authorize the app:
      // https://datahub.csail.mit.edu/permissions/apps/allow_access/APP_ID/REPO_NAME
      // see the /developer/apps page to register an application.

      String username = "foo_bar";
      String password = "foo_bar";
      String email = "albert.r.carter.foo_bar@gmail.com";
      String repo = "demo";       // the repository that your app operates on
      String table = "table"; // the table we will create
      String app_id = "test_id";             // your app's unique id
      String app_token = "e7633b8e-be1a-4bfa-9d02-b0611edfdfb3";       // your app's unique token
          

      // Before running you will need to populate the above variables
      // and create a corresponding repo and table in your datahub acount
          
      // http setup
      TTransport transport = new THttpClient("http://datahub.csail.mit.edu/service");
      TProtocol protocol = new  TBinaryProtocol(transport);
      AccountService.Client client = new AccountService.Client(protocol);
      System.out.println("Version: " + client.get_version());

      ConnectionParams con_params = new ConnectionParams();
      con_params.setUser(username);
      con_params.setApp_id(app_id);
      con_params.setApp_token(app_token);
      Connection con = client.open_connection(con_params);


      System.out.println("\n\nTrying to query a user's table...");
      String query = "select * from " + username + "." + repo + "." + table;
      ResultSet res = client.execute_sql(con, query, null);

      // print field names
      for (String field_name : res.getField_names()) {
        System.out.print(field_name + "\t");
      }

      System.out.println();

      // print tuples
      for (Tuple t : res.getTuples()) {
        for (ByteBuffer cell : t.getCells()) {
          System.out.print(new String(cell.array()) + "\t");
        }
        System.out.println();
      }
 
    } catch(Exception e) {
      e.printStackTrace();
    }
  }
}
